### ISSUE TITLE
"Navigation To" contains duplicated documents

### ISSUE THREAD

**Issue by vasily-kirichenko** [Microsoft/visualfsharp#2172]
![1](https://cloud.githubusercontent.com/assets/873919/21713178/a09f9190-d409-11e6-9aeb-b4fd1d9b529f.gif)

@dungpa any ideas?

**Comment by dungpa** [comment]
I saw it sometimes as well.

Maybe we haven't used `IDocumentNavigationService` in the right way: https://github.com/Microsoft/visualfsharp/blob/a24a5607861e3b546ef4bbeb28970eb4bbf0ecd1/vsintegration/src/FSharp.Editor/Navigation/NavigationBarItemService.fs#L77-L81?

**Comment by vasily-kirichenko** [comment]
Sorry, it was not Nav Bar, but Navigate To.

**Comment by vasily-kirichenko** [comment]
I've fixed the duplicates issue in https://github.com/Microsoft/visualfsharp/pull/2173

I've look at `Document.Id` of the returning items and the actual initially opened document, and they are identical. So, we have the following problem:

* The document which contains the searched symbol definition is opened.
* Navigate To this symbol.
* We pass proper item to Roslyn (Document + span).
* VS opens another copy of the document.
* It works OK for C# symbols.

**Comment by vasily-kirichenko** [comment]
@CyrusNajmabadi @Pilchie Any ideas what could, in theory, cause two copies of the same document to show at the same time, on different tabs? What are we missing?

**Comment by Pilchie** [comment]
I would try subscribing to `IVsRunningDocumentTableEvents` and seeing what the callstack is when the second document gets opened.  VS isn't *supposed* to let multiple copies of the same file get opened unless it is through Window.NewWindow, which would show a `:1` on the tab.

**Comment by vasily-kirichenko** [comment]
@Pilchie Done. I've added a listener:
```fsharp
let runningDocumentTableEvents =
    { new IVsRunningDocTableEvents with
        member __.OnAfterFirstDocumentLock(_, _, _, _) = VSConstants.S_OK
        member __.OnBeforeLastDocumentUnlock(_, _, _, _) = VSConstants.S_OK
        member __.OnAfterSave _ = VSConstants.S_OK
        member __.OnAfterAttributeChange(_, _) = VSConstants.S_OK
        member __.OnBeforeDocumentWindowShow(docCookie, fFirstShow, pFrame) = 
            VSConstants.S_OK
        member __.OnAfterDocumentWindowHide(_, _) = VSConstants.S_OK }

let vsRunningDocumentTable = Package.GetGlobalService(typeof<SVsRunningDocumentTable>) :?> IVsRunningDocumentTable
let sinkCookie = ref 0u
do vsRunningDocumentTable.AdviseRunningDocTableEvents(runningDocumentTableEvents, sinkCookie) |> ignore
```
Then invoked "Navigate To" to a symbol defined in a opened file. `OnBeforeDocumentWindowShow` was called 7 times (hm...), all the call stacks are here
https://gist.github.com/vasily-kirichenko/d9960d2f44b1a1db009941928a438e82

**Comment by Pilchie** [comment]
Take a look at `OnAfterFirstDocumentLock` when dwEditLocks + dwReadLocks = 1, and see what the other parameters are.

**Comment by vasily-kirichenko** [comment]
When the first copy is opening:

![image](https://cloud.githubusercontent.com/assets/873919/21725759/cee09e28-d44a-11e6-8152-e08cd07a5c6c.png)

It does not trigger _at all_ when the second copy is opening.

**Comment by Pilchie** [comment]
Hmm - not sure what might be going on.  Are you able to debug the Roslyn bits?  Can you see if there are any differences in [`TryGetFrame()`](http://source.roslyn.io/Microsoft.VisualStudio.LanguageServices/R/e00d49647b29002e.html)?

Beyond that - is the F# project system involved in either of the two calls invoked from that method?

**Comment by vasily-kirichenko** [comment]
No, I cannot debug Roslyn sources.

About F# project system, I think everything we do when a document is opening is in this method https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Common/LanguageService.fs#L318 I'm not sure we do everything right there.

**Comment by vasily-kirichenko** [comment]
Looks like this changes https://github.com/Microsoft/visualfsharp/pull/2185/files#diff-0c23e32ca44693769b36b39686377d92R471 has fixed this bug :) cooool.

**Comment by saul** [comment]
@vasily-kirichenko can you confirm this has been fixed by #2185 and if so close - thanks

**Comment by vasily-kirichenko** [comment]
Yes, I cannot reproduce it anymore.
